NC 25.0.1+Libresign V6.1.2 error contact admin (but the signed document is created)

Hi

NC 25.0.1 and Libresign v6.1.2, I try to sign a document and after introduce the password it shows a popup error "Internal Error. Contact with admin". Same occurs with NC24 and v.5.2.1 as described in another post.

The .signed document is created in the repository but without the QR code (it's just a copy of the original document).

Nextcloud.log only shows this:

{"reqId":"Pkg42JzATDLVEEHqQtqE","level":3,"time":"2022-11-08T14:22:23+00:00","remoteAddr":"192.168.100.20","user":"test","app":"libresign","method":"POST","url":"/index.php/apps/libresign/api/0.1/s
ign/uuid/9affedfe-803b-4791-b6e0-3e44c11e9b0b","message":"Error to sign PDF. [\"FINE Default property file doesn't exists.\",\"FINE Default property file doesn't exists.\",\"INFO Checking input and
output PDF paths.\",\"java.io.IOException: parseAlgParameters failed: ObjectIdentifier() -- data isn't an object ID (tag = 48)\",\"\\tat sun.security.pkcs12.PKCS12KeyStore.parseAlgParameters(PKCS12K
eyStore.java:787)\",\"\\tat sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:1951)\",\"\\tat java.security.KeyStore.load(KeyStore.java:1445)\",\"\\tat net.sf.jsignpdf.utils.KeyStore
Utils.loadKeyStore(KeyStoreUtils.java:359)\",\"\\tat net.sf.jsignpdf.utils.KeyStoreUtils.getPkInfo(KeyStoreUtils.java:411)\",\"\\tat net.sf.jsignpdf.SignerLogic.signFile(SignerLogic.java:154)\",\"\\
tat net.sf.jsignpdf.Signer.signFiles(Signer.java:246)\",\"\\tat net.sf.jsignpdf.Signer.main(Signer.java:139)\",\"Caused by: java.io.IOException: ObjectIdentifier() -- data isn't an object ID (tag =
48)\",\"\\tat sun.security.util.ObjectIdentifier.<init>(ObjectIdentifier.java:253)\",\"\\tat sun.security.util.DerInputStream.getOID(DerInputStream.java:281)\",\"\\tat com.sun.crypto.provider.PBES2P
arameters.engineInit(PBES2Parameters.java:267)\",\"\\tat java.security.AlgorithmParameters.init(AlgorithmParameters.java:293)\",\"\\tat sun.security.pkcs12.PKCS12KeyStore.parseAlgParameters(PKCS12Ke
yStore.java:783)\",\"\\t... 7 more\",\"WARNING Keystore was not loaded succesfully. Check if the keystore type, path and password are valid.\",\"SEVERE Problem occured\",\"java.lang.NullPointerExcep
tion: Keystore was not loaded succesfully. Check if the keystore type, path and password are valid.\",\"\\tat net.sf.jsignpdf.utils.KeyStoreUtils.getKeyAliasInternal(KeyStoreUtils.java:224)\",\"\\ta
t net.sf.jsignpdf.utils.KeyStoreUtils.getPkInfo(KeyStoreUtils.java:413)\",\"\\tat net.sf.jsignpdf.SignerLogic.signFile(SignerLogic.java:154)\",\"\\tat net.sf.jsignpdf.Signer.signFiles(Signer.java:24
6)\",\"\\tat net.sf.jsignpdf.Signer.main(Signer.java:139)\",\"\",\"INFO Finished: Creating of signature failed.\"]","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (K
HTML, like Gecko) Chrome/106.0.0.0 Safari/537.36","version":"25.0.1.1","data":{"app":"libresign"}}

Hi, sorry for the delay. Exactly the same error if I execute the command from the shell, it's in spanish, but basically the same error "WARNING Keystore was not loaded succesfully.":

DETALLADO Default property file doesn't exists.
DETALLADO Default property file doesn't exists.
INFORMACIÓN Comprobación de rutas de PDF de entrada y salida.
java.io.IOException: parseAlgParameters failed: ObjectIdentifier() -- data isn't an object ID (tag = 48)
	at sun.security.pkcs12.PKCS12KeyStore.parseAlgParameters(PKCS12KeyStore.java:787)
	at sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:1951)
	at java.security.KeyStore.load(KeyStore.java:1445)
	at net.sf.jsignpdf.utils.KeyStoreUtils.loadKeyStore(KeyStoreUtils.java:359)
	at net.sf.jsignpdf.utils.KeyStoreUtils.getPkInfo(KeyStoreUtils.java:411)
	at net.sf.jsignpdf.SignerLogic.signFile(SignerLogic.java:154)
	at net.sf.jsignpdf.Signer.signFiles(Signer.java:246)
	at net.sf.jsignpdf.Signer.main(Signer.java:139)
Caused by: java.io.IOException: ObjectIdentifier() -- data isn't an object ID (tag = 48)
	at sun.security.util.ObjectIdentifier.<init>(ObjectIdentifier.java:253)
	at sun.security.util.DerInputStream.getOID(DerInputStream.java:281)
	at com.sun.crypto.provider.PBES2Parameters.engineInit(PBES2Parameters.java:267)
	at java.security.AlgorithmParameters.init(AlgorithmParameters.java:293)
	at sun.security.pkcs12.PKCS12KeyStore.parseAlgParameters(PKCS12KeyStore.java:783)
	... 7 more
ADVERTENCIA El almacén de claves no se cargó correctamente. Compruebe si el tipo de almacén de claves, la ruta y la contraseña son válidos.
GRAVE Ocurrió un problema
java.lang.NullPointerException: El almacén de claves no se cargó correctamente. Compruebe si el tipo de almacén de claves, la ruta y la contraseña son válidos.
	at net.sf.jsignpdf.utils.KeyStoreUtils.getKeyAliasInternal(KeyStoreUtils.java:224)
	at net.sf.jsignpdf.utils.KeyStoreUtils.getPkInfo(KeyStoreUtils.java:413)
	at net.sf.jsignpdf.SignerLogic.signFile(SignerLogic.java:154)
	at net.sf.jsignpdf.Signer.signFiles(Signer.java:246)
	at net.sf.jsignpdf.Signer.main(Signer.java:139)

INFORMACIÓN Finalizado: No se pudo crear la firma.

--- a/lib/Service/ConfigureCheckService.php
+++ b/lib/Service/ConfigureCheckService.php
@@ -197,10 +197,10 @@ class ConfigureCheckService {
 					->setTip('Run occ libresign:install --java'),
 			];
 		}
-		$javaVersion = exec("java -version 2>&1");
+		\exec("java -version 2>&1", $javaVersion);
+		$javaVersion = current($javaVersion);
 		$hasJavaVersion = strpos($javaVersion, 'not found') === false;
 		if ($hasJavaVersion) {
-			$javaVersion = current($javaVersion);
 			return [
 				(new ConfigureCheckHelper())
 					->setSuccessMessage('Using java from operational system. Version: ' . $javaVersion)

--- a/lib/Service/ConfigureCheckService.php
+++ b/lib/Service/ConfigureCheckService.php
@@ -293,10 +293,22 @@ class ConfigureCheckService {
 			];
 		}
 		$return = [];
+		$version = str_replace("\n", ', ', trim(`$binary version`));
+		if (strpos($version, InstallService::CFSSL_VERSION) === false) {
+			return [
+				(new ConfigureCheckHelper())
+					->setErrorMessage(sprintf(
+						'Invalid version. Expected: %s, actual: %s',
+						InstallService::CFSSL_VERSION,
+						$version
+					))
+					->setResource('cfssl')
+					->setTip('Run occ libresign:install --cfssl')
+			];
+		}
 		$return[] = (new ConfigureCheckHelper())
 			->setSuccessMessage('CFSSL binary path: ' . $binary)
 			->setResource('cfssl');
-		$version = str_replace("\n", ', ', trim(`$binary version`));
 		$return[] = (new ConfigureCheckHelper())
 			->setSuccessMessage('CFSSL: ' . $version)
 			->setResource('cfssl');

--- a/lib/Service/ConfigureCheckService.php
+++ b/lib/Service/ConfigureCheckService.php
@@ -166,7 +166,21 @@ class ConfigureCheckService {
 		if ($javaPath) {
 			if (file_exists($javaPath)) {
 				\exec($javaPath . " -version 2>&1", $javaVersion);
-				$javaVersion = implode(\OC::$CLI ? "\n" : "<br>", $javaVersion);
+				$javaVersion = current($javaVersion);
+				if ($javaVersion !== InstallService::JAVA_VERSION) {
+					return [
+						(new ConfigureCheckHelper())
+							->setErrorMessage(
+								sprintf(
+									"Invalid java version. Found: %s expected: %s",
+									$javaVersion,
+									InstallService::JAVA_VERSION
+								)
+							)
+							->setResource('java')
+							->setTip('Run occ libresign:install --java'),
+					];
+				}
 				return [
 					(new ConfigureCheckHelper())
 						->setSuccessMessage('Java version: ' . $javaVersion)
@@ -186,6 +200,7 @@ class ConfigureCheckService {
 		$javaVersion = exec("java -version 2>&1");
 		$hasJavaVersion = strpos($javaVersion, 'not found') === false;
 		if ($hasJavaVersion) {
+			$javaVersion = current($javaVersion);
 			return [
 				(new ConfigureCheckHelper())
 					->setSuccessMessage('Using java from operational system. Version: ' . $javaVersion)

--- a/lib/Service/InstallService.php
+++ b/lib/Service/InstallService.php
@@ -25,6 +25,8 @@ use Symfony\Component\Console\Output\OutputInterface;
 use Symfony\Component\Process\Process;
 
 class InstallService {
+	public const JAVA_VERSION = 'openjdk version "17.0.5" 2022-10-18';
+	private const JAVA_PARTIAL_VERSION = '17.0.5_8';
 	/** @var ICache */
 	private $cache;
 	/** @var IConfig */
@@ -184,19 +186,20 @@ class InstallService {
 		}
 
 		/**
-		 * To update:
-		 * Check the compatible version of Java to use JSignPdf and update all the follow data
+		 * Steps to update:
+		 *     Check the compatible version of Java to use JSignPdf
+		 *     Update all the follow data
+		 *     Update the constants with java version
 		 * URL used to get the MD5 and URL to download:
 		 * https://jdk.java.net/java-se-ri/8-MR3
 		 */
 		if (PHP_OS_FAMILY === 'Linux') {
 			$architecture = php_uname('m');
-			$version = '17.0.5_8';
 			if ($architecture === 'x86_64') {
-				$compressedFileName = 'OpenJDK17U-jre_x64_linux_hotspot_' . $version . '.tar.gz';
+				$compressedFileName = 'OpenJDK17U-jre_x64_linux_hotspot_' . self::JAVA_PARTIAL_VERSION . '.tar.gz';
 				$url = 'https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.5%2B8/' . $compressedFileName;
 			} elseif ($architecture === 'aarch64') {
-				$compressedFileName = 'OpenJDK17U-jre_aarch64_linux_hotspot_' . $version . '.tar.gz';
+				$compressedFileName = 'OpenJDK17U-jre_aarch64_linux_hotspot_' . self::JAVA_PARTIAL_VERSION . '.tar.gz';
 				$url = 'https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.5%2B8/' . $compressedFileName;
 			}
 			$class = TAR::class;
@@ -205,7 +208,7 @@ class InstallService {
 		}
 		$folder = $this->getFolder();
 		$checksumUrl = $url . '.sha256.txt';
-		$hash = $this->getHash($folder, 'java', $compressedFileName, $version, $checksumUrl);
+		$hash = $this->getHash($folder, 'java', $compressedFileName, self::JAVA_PARTIAL_VERSION, $checksumUrl);
 		if (!$javaFolder->nodeExists($compressedFileName)) {
 			$compressedFile = $javaFolder->newFile($compressedFileName);
 		} else {

--- a/lib/Service/InstallService.php
+++ b/lib/Service/InstallService.php
@@ -27,6 +27,10 @@ use Symfony\Component\Process\Process;
 class InstallService {
 	public const JAVA_VERSION = 'openjdk version "17.0.5" 2022-10-18';
 	private const JAVA_PARTIAL_VERSION = '17.0.5_8';
+	/**
+	 * When update, verify the hash of all architectures
+	 */
+	public const CFSSL_VERSION = '1.6.3';
 	/** @var ICache */
 	private $cache;
 	/** @var IConfig */
@@ -373,22 +377,21 @@ class InstallService {
 
 	private function installCfssl64(): void {
 		$folder = $this->getFolder();
-		$version = '1.6.1';
 
 		$downloads = [
 			[
-				'file' => 'cfssl_' . $version . '_linux_amd64',
+				'file' => 'cfssl_' . self::CFSSL_VERSION . '_linux_amd64',
 				'destination' => 'cfssl',
 			],
 			[
-				'file' => 'cfssljson_' . $version . '_linux_amd64',
+				'file' => 'cfssljson_' . self::CFSSL_VERSION . '_linux_amd64',
 				'destination' => 'cfssljson',
 			],
 		];
-		$baseUrl = 'https://github.com/cloudflare/cfssl/releases/download/v' . $version . '/';
-		$checksumUrl = 'https://github.com/cloudflare/cfssl/releases/download/v' . $version . '/cfssl_' . $version . '_checksums.txt';
+		$baseUrl = 'https://github.com/cloudflare/cfssl/releases/download/v' . self::CFSSL_VERSION . '/';
+		$checksumUrl = 'https://github.com/cloudflare/cfssl/releases/download/v' . self::CFSSL_VERSION . '/cfssl_' . self::CFSSL_VERSION . '_checksums.txt';
 		foreach ($downloads as $download) {
-			$hash = $this->getHash($folder, 'libresign-cli', $download['file'], $version, $checksumUrl);
+			$hash = $this->getHash($folder, 'libresign-cli', $download['file'], self::CFSSL_VERSION, $checksumUrl);
 
 			$file = $folder->newFile($download['destination']);
 			$fullPath = $this->getDataDir() . DIRECTORY_SEPARATOR . $file->getInternalPath();
@@ -412,7 +415,7 @@ class InstallService {
 		} else {
 			$cfsslFolder = $appFolder->newFolder('cfssl');
 		}
-		$compressedFileName = 'cfssl-1.6.3-1-aarch64.pkg.tar.xz';
+		$compressedFileName = 'cfssl-' . self::CFSSL_VERSION . '-1-aarch64.pkg.tar.xz';
 		$url = 'http://mirror.archlinuxarm.org/aarch64/community/' . $compressedFileName;
 		// Generated handmade with command sha256sum
 		$hash = '944a6c54e53b0e2ef04c9b22477eb5f637715271c74ccea9bb91d7ac0473b855';

--- a/package.json
+++ b/package.json
@@ -25,7 +25,7 @@
     "@nextcloud/password-confirmation": "^4.0.2",
     "@nextcloud/paths": "^2.1.0",
     "@nextcloud/router": "^2.0.0",
-    "@nextcloud/vue": "^7.2.0",
+    "@nextcloud/vue": "^7.3.0",
     "crypto-js": "^4.1.1",
     "date-fns": "^2.29.1",
     "dompurify": "^2.4.1",

--- a/package.json
+++ b/package.json
@@ -29,6 +29,7 @@
     "crypto-js": "^4.1.1",
     "date-fns": "^2.29.1",
     "dompurify": "^2.4.1",
+    "linkify-string": "^4.0.2",
     "lodash-es": "^4.17.21",
     "marked": "^4.2.4",
     "style-loader": "^3.3.1",

